yoyodyne/predict.py

```diff
@@ -1,50 +1,142 @@
 """Prediction."""
 
+import argparse
 import os
-from typing import Optional
 
-import click
 import pytorch_lightning as pl
 import torch
 from torch.utils import data
 
 from . import collators, dataconfig, datasets, models, util
 
 
-def write_predictions(
-    model: models.base.BaseEncoderDecoder,
-    loader: torch.utils.data.DataLoader,
-    output: str,
-    accelerator: str,
+def get_trainer(**kwargs) -> pl.Trainer:
+    """Creates the trainer.
+
+    Args:
+        **kwargs: passed to the trainer.
+
+    Returns:
+        pl.Trainer.
+    """
+    return pl.Trainer(max_epochs=0, **kwargs)
+
+
+def _get_trainer_from_argparse_args(
+    args: argparse.Namespace,
+) -> pl.Trainer:
+    """Creates the trainer from CLI arguments."""
+    return pl.Trainer.from_argparse_args(args, max_epochs=0)
+
+
+def get_dataset(
+    predict: str,
+    config: dataconfig.DataConfig,
+    index: str,
+) -> data.Dataset:
+    """Creates the dataset.
+
+    Args:
+        predict (str).
+        config (dataconfig.DataConfig).
+        index (str).
+
+    Returns:
+        data.Dataset.
+    """
+    # TODO: Since we don't care about the target column, we should be able to
+    # set config.source_col = 0 and avoid the overhead for parsing it.
+    # This does not work because the modules expect it to be present even if
+    # they ignore it.
+    dataset = datasets.get_dataset(predict, config)
+    dataset.read_index(index)
+    return dataset
+
+
+def get_loader(
+    dataset: data.Dataset,
+    arch: str,
     batch_size: int,
-    target_sep: str = "",
-    beam_width: Optional[int] = None,
-) -> None:
-    """Writes predictions to output file.
+) -> data.DataLoader:
+    """Creates the loader.
 
     Args:
-        model (models.BaseEncoderDecoder).
-        loader (torch.utils.data.DataLoader).
-        output (str).
+        dataset (data.Dataset).
         arch (str).
-        accelerator (str).
         batch_size (int).
-        target_sep (str).
-        beam_width (int, optional).
+
+    Returns:
+        data.DataLoader.
     """
-    model.beam_width = beam_width
-    model.eval()
-    # Test loop.
-    tester = pl.Trainer(
-        accelerator=accelerator,
-        max_epochs=0,  # Silences a warning.
+    collator = collators.Collator(dataset.pad_idx, dataset.config, arch)
+    return data.DataLoader(
+        dataset,
+        collate_fn=collator,
+        batch_size=batch_size,
+        num_workers=1,  # Our data loading is simple.
     )
-    util.log_info("Predicting...")
+
+
+def _get_loader_from_argparse_args(
+    args: argparse.Namespace,
+) -> data.DataLoader:
+    """Creates the loader from CLI arguments."""
+    config = dataconfig.DataConfig.from_argparse_args(args)
+    dataset = get_dataset(args.predict, config, args.index)
+    return get_loader(dataset, args.arch, args.batch_size)
+
+
+def get_model(
```

Comment:
What is this for?

Reply:
This and `get_trainer` are reasonably needed around if you want to use this tool not-from-the-CLI, I hypothesize. I may make small changes to them when I start writing deeper tests.

Reply:
Good point I forgot about that use case ha